webpack.config.js

```diff
@@ -21,46 +20,204 @@ module.exports = {
+        // the defaults of webpack prepended with `matrix_src_`.
+        mainFields: ['matrix_src_browser', 'matrix_src_main', 'browser', 'main'],
+        aliasFields: ['matrix_src_browser', 'browser'],
+
+        // We need to ensure we can resolve TS files, but that also means we need to define
+        // every single extension we might see, ever.
+        extensions: [
+            '.js',
+            '.json',
+            '.css',
+            '.scss',
+            '.ts',
+            '.gif',
+            '.png',
+            '.svg',
+            '.ttf',
+            '.woff',
+            '.woff2',
+            '.xml',
+            '.ico',
+            '.wasm',
+        ],
+        alias: {
+            // alias any requires to the react module to the one in our path,
+            // otherwise we tend to get the react source included twice when
+            // using `npm link` / `yarn link`.
+            "react": path.resolve(__dirname, 'node_modules/react'),
+            "react-dom": path.resolve(__dirname, 'node_modules/react-dom'),
+
+            // same goes for js-sdk - we don't need two copies.
+            "matrix-js-sdk": path.resolve(__dirname, 'node_modules/matrix-js-sdk'),
+
+            // Define a variable so the i18n stuff can load
+            "$webapp": path.resolve(__dirname, 'webapp'),
+        },
+    },
+
     module: {
+        noParse: [
+            // for cross platform compatibility use [\\\/] as the path separator
+            // this ensures that the regex trips on both Windows and *nix
+
+            // don't parse the languages within highlight.js. They cause stack
+            // overflows (https://github.com/webpack/webpack/issues/1721), and
+            // there is no need for webpack to parse them - they can just be
+            // included as-is.
+            /highlight\.js[\\\/]lib[\\\/]languages/,
+
+            // olm takes ages for webpack to process, and it's already heavily
+            // optimised, so there is little to gain by us uglifying it.
+            /olm[\\\/](javascript[\\\/])?olm\.js$/,
+        ],
         rules: [
-            { enforce: 'pre', test: /\.js$/, use: "source-map-loader", exclude: /node_modules/, },
-            { test: /\.js$/, use: "babel-loader", include: path.resolve(__dirname, 'src') },
             {
-                test: /\.wasm$/,
-                loader: "file-loader",
-                type: "javascript/auto", // https://github.com/webpack/webpack/issues/6725
+                test: /\.(ts|js)x?$/,
+                exclude: /node_modules/,
+                loader: 'babel-loader',
                 options: {
-                    name: '[name].[hash:7].[ext]',
-                    outputPath: '.',
-                },
+                    // We have the babel config here rather than a file to help
+                    // make it clearer what we're compiling to.
+                    cacheDirectory: true,
+                    sourceMaps: true,
```

Comment:
This seems surprising: wouldn't this be for if you're running babel separately? In this case shouldn't webpack being telling its loaders to give it whatever info it needs to make a source map?

edit: oh, having read the whole file I see the other comment. This is surprising and slightly terrifying. How did you discover that worked?

Reply:
Webpack's sourcemaps are the problem and why they're off by 1 line, so we grab babel's sourcemaps which are perfect. Digging into the oddities of how webpack works to figure out why they're off by 1 line, even after changing sourcemap plugins, is not something that's worth our time at the moment I think.

Reply:
Mmm, agreed - I'm just surprised that this does anything as I would expect this would make babel-loader produce source maps for each file and then hand them to webpack which would do nothing with them because it's been told not to make a source map, so webpack (or something) must be taking all those individual ones and turning them into a source map for the bundle. If it somehow works then... yay, I guess.

Reply:
yea, not totally sure what's going on to make it work the way it does, but there's a lot of goat's blood on the table.

Reply:
🐐